### ISSUE TITLE
gravitational potential is hard-coded to be a function of radius

### ISSUE THREAD

**Issue by BenWibking** [parthenon-hpc-lab/athenapk#12]
It looks like the gravitational potential is hard-coded to be only a function of radius. In general, it should be a function of all of the coordinates. (For my simulations, it should be a function of height.)

**Comment by BenWibking** [comment]
It's a simple function. Should it go in `PostStepMeshUserWorkInLoop`?

**Comment by pgrete** [comment]
I suggest to hook into the `Hydro::ProblemSourceUnsplit` or `Hydro::ProblemSourceFirstOrder` function (cf., the turbulence problem generator).
The "unplit" version is called during every stage of the integrator (i..e, the `beta_dt` should be used for $\Delta t$) whereas the "first order" one is only called once (so the full $\Delta t$ for the entire cycle should be used).

**Comment by pgrete** [comment]
> Is there a Strang-split version of this? Otherwise I'll just use `Hydro::ProblemSourceFirstOrder`.

Yes, it's called `ProblemSourceStrangSplit` :smile: 

Also this applies: https://github.com/parthenon-hpc-lab/athenapk/blob/main/src/hydro/hydro_driver.cpp#L160

**Comment by BenWibking** [comment]
Well, that's convenient. So I need to read from the cons variables and then update both cons and prim?

**Comment by BenWibking** [comment]
What is the reason for this restriction ? Could this be simplified by adding another conserved -> primitive conversion step in the task list?

**Comment by pgrete** [comment]
> Well, that's convenient. So I need to read from the cons variables and then update both cons and prim?

yes (that's the "safe" option)

> What is the reason for this restriction ? Could this be simplified by adding another conserved -> primitive conversion step in the task list?

The current implementation is mostly motivated by performance concerns, which doesn't mean it has to stay that way.
So far the source terms I've been using update the data they were reading, so I could easily update both prim and cons in those source terms without needing to read and write all data again.
For example,  for cooling the update to the pressure is trivial based on required calculations in the kernel whereas a separate ConsToPrim would need to read all data (density, momentum, energy, and potentially magnetic fields) again, just to update the pressure variable.

**Comment by BenWibking** [comment]
Ok, that makes sense. I agree about there being a performance cost, but I'd expect it to be very small - have you measured this?

**Comment by pgrete** [comment]
> Ok, that makes sense. I agree about there being a performance cost, but I'd expect it to be very small - have you measured this?

Yes, (but it also strongly depends on the total amount of physics included).
IIRC it's double digit percent for MHD because that kernel is extremely memory bound as there virtual no compute for cons to prim conversion.

**Comment by BenWibking** [comment]
I am getting some exceptionally weird (wrong) results when using Strang splitting for gravity:

Working version (first order splitting):
![parthenon prim final_Slice_x_Velocity3](https://user-images.githubusercontent.com/1781919/214131839-98a9c0b3-cde6-432d-b226-febfd189883c.png)

Bad version (Strang splitting):
![parthenon prim final_Slice_x_Velocity3_strang](https://user-images.githubusercontent.com/1781919/214131883-28b85ecc-a7a0-46a6-985b-7a13682216e1.png)

Do you have any idea what could be going wrong here?

**Comment by pgrete** [comment]
That does indeed look strange.
If you could provide a reproducer (i.e., the code version/commit to use an a sample input file), I'll have a look.
